packages/@aws-cdk/aws-servicecatalog/lib/product.ts

```diff
@@ -202,20 +214,46 @@ export class CloudFormationProduct extends Product {
     }
   }
 
-  private renderProvisioningArtifacts(
-    props: CloudFormationProductProps): CfnCloudFormationProduct.ProvisioningArtifactPropertiesProperty[] {
-    return props.productVersions.map(productVersion => {
+  private renderProvisioningArtifacts(props: CloudFormationProductProps): CfnCloudFormationProduct.ProvisioningArtifactPropertiesProperty[] {
+    let productVersions: CfnCloudFormationProduct.ProvisioningArtifactPropertiesProperty[] = [];
+    for (const productVersion of props.productVersions) {
       const template = productVersion.cloudFormationTemplate.bind(this);
-      InputValidator.validateUrl(this.node.path, 'provisioning template url', template.httpUrl);
-      return {
-        name: productVersion.productVersionName,
-        description: productVersion.description,
-        disableTemplateValidation: productVersion.validateTemplate === false ? true : false,
-        info: {
-          LoadTemplateFromURL: template.httpUrl,
+      let httpUrl = template.httpUrl;
+      switch (template.templateType) {
+        case TemplateType.PRODUCT_STACK:
+          if (template.productVersionDetails) {
+            template.productVersionDetails.productPathUniqueId = this.productPathUniqueId;
+            template.productVersionDetails.productVersionName = productVersion.productVersionName;
+            template.productVersionDetails.versioningStrategy = productVersion.versioningStrategy;
+          }
+          break;
+        case TemplateType.PRODUCT_STACK_CONTEXT:
+          const templateFileKey = `${this.productPathUniqueId}.${template.productVersionDetails?.productStackId}.${productVersion.productVersionName}.product.template.json`;
```

Comment:
what happens if `template.productVersionDetails` is `undefined` here?

Reply:
ahh okay, this makes sense. Thanks for clarifying. @rix0rrr should we use `instanceof` here instead? That would remove the `?` operator here.

Reply:
Never `instanceof`! If you see `instanceof` it's a bug.

Reply:
@rix0rrr is that because inheritance should be used instead? Could you elaborate?